File: ios/ReactTestApp/Session.swift
//
// Copyright (c) Microsoft Corporation. All rights reserved.
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the root directory of this source tree.
//

import Foundation

struct Session {
    private enum Keys {
        static let checksum = "ManifestChecksum"
        static let rememberLastComponentEnabled = "RememberLastComponent/Enabled"
        static let rememberLastComponentIndex = "RememberLastComponent/Index"
    }

    private static var lastComponentIndex: Int {
        get { UserDefaults.standard.integer(forKey: Keys.rememberLastComponentIndex) }
        set { UserDefaults.standard.set(newValue, forKey: Keys.rememberLastComponentIndex) }
    }

    private static var manifestChecksum: String? {
        get { UserDefaults.standard.string(forKey: Keys.checksum) }
        set { UserDefaults.standard.set(newValue, forKey: Keys.checksum) }
    }

    static var shouldRememberLastComponent: Bool {
        get { UserDefaults.standard.bool(forKey: Keys.rememberLastComponentEnabled) }
        set { UserDefaults.standard.set(newValue, forKey: Keys.rememberLastComponentEnabled) }
    }

    static func lastOpenedComponent(_ checksum: String) -> Int? {
        guard shouldRememberLastComponent, checksum == manifestChecksum else {
            return nil
        }

        return lastComponentIndex
    }

    static func storeComponent(index: Int, checksum: String) {
        lastComponentIndex = index
        manifestChecksum = checksum
    }
}

File: android/app/src/main/java/com/microsoft/reacttestapp/Session.kt
package com.microsoft.reacttestapp

import android.content.Context
import android.content.Context.MODE_PRIVATE
import android.content.SharedPreferences
import androidx.core.content.edit

private const val MANIFEST_CHECKSUM = "ManifestChecksum"
private const val REMEMBER_LAST_COMPONENT_ENABLED = "RememberLastComponent/Enabled"
private const val REMEMBER_LAST_COMPONENT_INDEX = "RememberLastComponent/Index"

class Session(context: Context) {

    private val sharedPreferences: SharedPreferences =
        context.getSharedPreferences(BuildConfig.APPLICATION_ID, MODE_PRIVATE)

    private var lastComponentIndex: Int
        get() = sharedPreferences.getInt(REMEMBER_LAST_COMPONENT_INDEX, -1)
        set(index) = sharedPreferences.edit { putInt(REMEMBER_LAST_COMPONENT_INDEX, index) }

    private var manifestChecksum: String?
        get() = sharedPreferences.getString(MANIFEST_CHECKSUM, null)
        set(checksum) = sharedPreferences.edit { putString(MANIFEST_CHECKSUM, checksum) }

    var shouldRememberLastComponent: Boolean
        get() = sharedPreferences.getBoolean(REMEMBER_LAST_COMPONENT_ENABLED, false)
        set(enabled) = sharedPreferences.edit {
            putBoolean(REMEMBER_LAST_COMPONENT_ENABLED, enabled)
        }

    fun lastOpenedComponent(checksum: String): Int? {
        if (!shouldRememberLastComponent || checksum != manifestChecksum) {
            return null
        }

        val index = lastComponentIndex
        return if (index < 0) null else index
    }

    fun storeComponent(index: Int, checksum: String) {
        lastComponentIndex = index
        manifestChecksum = checksum
    }
}
